### ISSUE TITLE
Belos on gcc/opt/openmp failes on ride

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#1191]
**Next Action Status:**

LAPACK calling problems may not have been resolved on 'ride' with the merge of #1284. Next: @kruger is digging in more in #1208 ...

**Blocked By:** #1208 

**Description:**

Has anyone ran the tests for Belos on Ride with gcc?  I'm seeing these errors, and what looks like related issues 
174:Belos_pseudo_gmres_hb_MPI_4
175:Belos_bl_pgmres_hb_0_MPI_4
176:Belos_bl_pgmres_hb_1_MPI_4
177:Belos_bl_pgmres_hb_2_MPI_4
178:Belos_pseudo_pgmres_hb_MPI_4
179:Belos_bl_fgmres_hb_MPI_4
192:Belos_gcrodr_hb_MPI_4
194:Belos_prec_gcrodr_hb_1_MPI_4
218:Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4
237:Stratimikos_Thyra_Belos_StatusTest_UnitTests_MPI_1
245:Stratimikos_Belos_GCRODR_strattest_MPI_4

The tests seem to pass in debug mode.
@nmhamster have you seen any tests pass on ride?

@kddevin are you the PI for solvers like Belos?  Are you the POC for this?

**Comment by bathmatt** [comment]
Ok, there is one failure under debug

```
The following tests FAILED:
	 54 - Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4 (Failed)
Errors while running CTest
mbetten@ride11:~/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos$ more Testing/Temporary/LastTest.log 
Start testing: Mar 28 15:01 MDT
----------------------------------------------------------
54/59 Testing: Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4
54/59 Test: Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4
Command: "/ascldap/users/projects/pwr8-rhel73-lsf/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/bin/mpiexec" "-np" "4" "-map-by" "socket:PE=8" "--oversubscribe" "/ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetr
a/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe"
Directory: /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG
"Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4" start time: Mar 28 15:01 MDT
Output:
----------------------------------------------------------
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ride11 and rank 1!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ride11 and rank 3!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ride11 and rank 2!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name ride11 and rank 0!
Kokkos::OpenMP::initialize WARNING: You are likely oversubscribing your CPU cores.
                                    Detected: 128 cores per node.
                                    Detected: 4 MPI_ranks per node.
                                    Requested: 64 threads per process.
Kokkos::OpenMP::initialize WARNING: You are likely oversubscribing your CPU cores.
                                    Detected: 128 cores per node.
                                    Detected: 4 MPI_ranks per node.
                                    Requested: 64 threads per process.
Kokkos::OpenMP::initialize WARNING: You are likely oversubscribing your CPU cores.
                                    Detected: 128 cores per node.
                                    Detected: 4 MPI_ranks per node.
                                    Requested: 64 threads per process.
Kokkos::OpenMP::initialize WARNING: You are likely oversubscribing your CPU cores.
                                    Detected: 128 cores per node.
                                    Detected: 4 MPI_ranks per node.
                                    Requested: 64 threads per process.
 ** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value
[ride11:05891] *** Process received signal ***
[ride11:05891] Signal: Segmentation fault (11)
[ride11:05891] Signal code: Invalid permissions (2)
[ride11:05891] Failing at address: 0x3ffff96bd190
[ride11:05891] [ 0] [0x100000050478]
[ride11:05891] [ 1] [0x3ffff96bd190]
[ride11:05891] [ 2] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(_ZN5Belos19PseudoBlockCGSolMgrIdN6Tpetra11MultiVectorIdiiN6Kokkos6Compat23Kokko
sDeviceWrapperNodeINS3_6OpenMPENS3_9HostSpaceEEELb0EEENS1_8OperatorIdiiS8_EELb1EE5solveEv+0x15d0)[0x11234a24]
[ride11:05891] [ 3] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(main+0x1258)[0x1121a8d8]
[ride11:05891] [ 4] /lib64/libc.so.6(+0x24700)[0x100000854700]
[ride11:05891] [ 5] /lib64/libc.so.6(__libc_start_main+0xc4)[0x1000008548f4]
[ride11:05891] *** End of error message ***
[ride11:05892] *** Process received signal ***
[ride11:05892] Signal: Segmentation fault (11)
[ride11:05892] Signal code: Invalid permissions (2)
[ride11:05892] Failing at address: 0x3ffff3025e70
[ride11:05892] [ 0] [0x100000050478]
[ride11:05892] [ 1] [0x3ffff3025e70]
[ride11:05892] [ 2] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(_ZN5Belos19PseudoBlockCGSolMgrIdN6Tpetra11MultiVectorIdiiN6Kokkos6Compat23Kokko
sDeviceWrapperNodeINS3_6OpenMPENS3_9HostSpaceEEELb0EEENS1_8OperatorIdiiS8_EELb1EE5solveEv+0x15d0)[0x11234a24]
[ride11:05892] [ 3] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(main+0x1258)[0x1121a8d8]
[ride11:05892] [ 4] /lib64/libc.so.6(+0x24700)[0x100000854700]
[ride11:05892] [ 5] /lib64/libc.so.6(__libc_start_main+0xc4)[0x1000008548f4]
[ride11:05892] *** End of error message ***
[ride11:05893] *** Process received signal ***
[ride11:05893] Signal: Segmentation fault (11)
[ride11:05893] Signal code: Invalid permissions (2)
[ride11:05893] Failing at address: 0x3fffdb30e780
[ride11:05893] [ 0] [0x100000050478]
[ride11:05893] [ 1] [0x3fffdb30e780]
[ride11:05893] [ 2] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(_ZN5Belos19PseudoBlockCGSolMgrIdN6Tpetra11MultiVectorIdiiN6Kokkos6Compat23Kokko
sDeviceWrapperNodeINS3_6OpenMPENS3_9HostSpaceEEELb0EEENS1_8OperatorIdiiS8_EELb1EE5solveEv+0x15d0)[0x11234a24]
[ride11:05893] [ 3] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(main+0x1258)[0x1121a8d8]
[ride11:05893] [ 4] /lib64/libc.so.6(+0x24700)[0x100000854700]
[ride11:05893] [ 5] /lib64/libc.so.6(__libc_start_main+0xc4)[0x1000008548f4]
[ride11:05893] *** End of error message ***
[ride11:05894] *** Process received signal ***
[ride11:05894] Signal: Segmentation fault (11)
[ride11:05894] Signal code: Invalid permissions (2)
[ride11:05894] Failing at address: 0x3fffd3468e10
[ride11:05894] [ 0] [0x100000050478]
[ride11:05894] [ 1] [0x3fffd3468e10]
[ride11:05894] [ 2] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(_ZN5Belos19PseudoBlockCGSolMgrIdN6Tpetra11MultiVectorIdiiN6Kokkos6Compat23Kokko
sDeviceWrapperNodeINS3_6OpenMPENS3_9HostSpaceEEELb0EEENS1_8OperatorIdiiS8_EELb1EE5solveEv+0x15d0)[0x11234a24]
[ride11:05894] [ 3] /ascldap/users/mbetten/workspace/build_unknown_GNU_DEBUG_OPENMP/packages/belos/tpetra/test/BlockCG/Belos_Tpetra_PseudoBlockCG_hb_test.exe(main+0x1258)[0x1121a8d8]
[ride11:05894] [ 4] /lib64/libc.so.6(+0x24700)[0x100000854700]
[ride11:05894] [ 5] /lib64/libc.so.6(__libc_start_main+0xc4)[0x1000008548f4]
[ride11:05894] *** End of error message ***
--------------------------------------------------------------------------
mpiexec noticed that process rank 3 with PID 5894 on node ride11 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
<end of output>
Test time = 176.24 sec
----------------------------------------------------------
Test Failed.
"Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4" end time: Mar 28 15:04 MDT
"Belos_Tpetra_PseudoBlockCG_hb_test_MPI_4" time elapsed: 00:02:56
----------------------------------------------------------

End testing: Mar 28 15:04 MDT

Belos = 176.24 sec



```

**Comment by mhoemmen** [comment]
Hi @bathmatt !  Thanks for reporting!  Some questions:

  - Are these tests with or without CUDA?
  - Have you set OMP_NUM_THREADS so that OpenMP doesn't oversubscribe?
  - What GCC and MPI versions are you using?

**Comment by kddevin** [comment]
Hi, @bathmatt .  No, I am not the PI for Belos.  I'll tag the Belos team @trilinos/belos and @hkthorn .
@mhoemmen  is already following.

**Comment by bathmatt** [comment]
@mhoemmen <https://github.com/mhoemmen>
OMP_NUM_THREADS=4

This is the same configure that works for panzer module a couple solver
issues

mbetten@ride6:~/workspace/build_unknown_GNU_RELEASE_OPENMP$ module list
Currently Loaded Modulefiles:
  1) seacas/20170224/openmpi/none/gcc/5.4.0/cuda/none      9)
isl/0.17.1                                           17)
cgns/3.2.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44      25) cmake/3.6.2
  2) cuda/8.0.44                                          10)
cloog/0.18.4                                         18)
openblas/0.2.19/gcc/5.4.0                            26) tcl/8.5.19
  3) numa/2.0.11                                          11)
gcc/5.4.0                                            19)
superlu/4.3.0/gcc/5.4.0                              27) perl/5.22.1
  4) papi/5.5.0                                           12)
java/ibm/sdk/8.0.0                                   20)
boost/1.60.0/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44    28) openssl/1.0.2j
  5) zlib/1.2.8                                           13)
openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44                 21)
metis/5.0.1/gcc/5.4.0                                29) git/2.10.1
  6) mpfr/3.1.5                                           14)
pnetcdf/1.6.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44   22)
parmetis/4.0.3/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44  30) valgrind/3.11.0
  7) mpc/1.0.3                                            15)
hdf5/1.8.17/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44     23)
yaml-cpp/20170104                                    31)
devpack/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
  8) gmp/6.1.0                                            16)
netcdf/4.4.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44    24) curl/7.47.1

**Comment by mhoemmen** [comment]
@bathmatt Could you post your Trilinos configuration script here please?

**Comment by mhoemmen** [comment]
btw here is a possible valid order for loading the modules on ride:
```
module load git/2.10.1
module load cmake/3.6.2
module load tcl/8.5.19
module load perl/5.22.1
module load numa/2.0.11
module load papi/5.5.0
module load isl/0.17.1
module load cloog/0.18.4
module load gmp/6.1.0
module load mpfr/3.1.5
module load mpc/1.0.3
module load gcc/5.4.0
module load cuda/8.0.44
module load openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load openblas/0.2.19/gcc/5.4.0
module load boost/1.60.0/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load superlu/4.3.0/gcc/5.4.0
module load cgns/3.2.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load seacas/20170224/openmpi/none/gcc/5.4.0/cuda/none
module load java/ibm/sdk/8.0.0
module load openssl/1.0.2j
module load zlib/1.2.8
module load metis/5.0.1/gcc/5.4.0
module load pnetcdf/1.6.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load parmetis/4.0.3/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load valgrind/3.11.0
module load hdf5/1.8.17/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load devpack/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load netcdf/4.4.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
module load yaml-cpp/20170104
module load curl/7.47.1
```
There was a little bit of weirdness regarding yaml-cpp: the load order mattered, and `yaml-cpp/20170104` appears to duplicate the `yamlcpp/0.3.0` module that a previous module loaded.

**Comment by hkthorn** [comment]
The error is:

** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value
 ** On entry to DSTEQR parameter number  1 had an illegal value

This leads to the condition number estimate calculation code, more specifically the first argument of that call which is a character.  Sometimes the argument lists for LAPACK that use characters have to be modified due to the platform.  Thus the CHAR_MACRO that is in the Teuchos LAPACK interface.  It may be a red herring, but that is where the error above leads.

**Comment by bathmatt** [comment]
[config.txt](https://github.com/trilinos/Trilinos/files/880468/config.txt)
THis is a tar file really, not a txt..  Also, set JOB_NAME ride_gnu_opt_openmp and workspace to where you have trilinos, and remove all the drekar lines and you should be good to go

**Comment by mhoemmen** [comment]
I've been too busy lately but I can pick this up.

**Comment by bathmatt** [comment]
Thanks Mark!!

**Comment by mhoemmen** [comment]
Updating my Trilinos:develop on ride now.  Will try out your config script once that's done.

**Comment by bathmatt** [comment]
Recall,
This is optimized only that fails.

**Comment by mhoemmen** [comment]
@bathmatt so if I change your build script to a RELEASE build, will that manifest the issue?

**Comment by mhoemmen** [comment]
@bathmatt How do I use the two scripts?  What do I need to change in each of them?

**Comment by bathmatt** [comment]
from previous comment
THis is a tar file really, not a txt.. Also, set JOB_NAME ride_gnu_opt_openmp and workspace to where you have trilinos, and remove all the drekar lines and you should be good to go

**Comment by bathmatt** [comment]
[out.txt](https://github.com/trilinos/Trilinos/files/893956/out.txt)
should be self contained. haven't tested it

**Comment by mhoemmen** [comment]
@bathmatt Trying this out now -- it's working thus far (mid-way through Trilinos configure process).  Thanks!

**Comment by mhoemmen** [comment]
Building Trilinos now.

**Comment by bathmatt** [comment]
This is my run script for the build and test.  the CUDA ones are ignored
for non-cuda builds

```

cat > job.sh << "EOF"
#! /bin/bash
#PBS -l nodes=1:ppn=64
#PBS -l cput=06:00:00
#PBS -N $JOB_NAME

. ./Trilinos/DrekarBase/drekar/build_scripts/ride/environment.sh
grep -v configure
./Trilinos/DrekarBase/drekar/build_scripts/ride/environment.sh
>>$INSTALL_DIR/configure.sh
env
. ./Trilinos/DrekarBase/drekar/build_scripts/ALL/build_drekar.sh
cd $WORKSPACE/${BUILD_DIR}
export CUDA_LAUNCH_BLOCKING=1
export CUDA_MANAGED_FORCE_DEVICE_ALLOC=1
export OMP_NUM_THREADS=4
ulimit -c 0
ctest -j16 -T test
EOF
```

**Comment by mhoemmen** [comment]
@hkthorn @bathmatt I think I get it.  In the reference LAPACK implementation, DSTEQR prints that message when it returns `INFO = -1`.  This means that the first argument, the character array indicating what to compute, is neither "N" or "V" nor "I".

This smells of a Fortran binding issue for strings.  Some Fortran compilers have an ABI that passes character arrays as `char*, int`.  Others just pass them as `char*`.  I think there is a quick way to test this in Teuchos; I'll try it.

**Comment by mhoemmen** [comment]
I'm testing this theory now.

**Comment by mhoemmen** [comment]
Hm, that didn't seem to help.  Let me try hard-coding "N" in Teuchos::LAPACK's implementation.

**Comment by mhoemmen** [comment]
Oddly enough, hard-coding "N" in Teuchos::LAPACK's implementation worked...

**Comment by mhoemmen** [comment]
In summary: If I change lines 735-738 of `teuchos/numerics/src/Teuchos_LAPACK.cpp` to the following:
```
  void LAPACK<int,double>::STEQR(const char COMPZ, const int n, double* D, double* E, double* Z, const int ldz, double* WORK, int* info) const
  {
    DSTEQR_F77("N", &n, D, E, Z, &ldz, WORK, info);
  }
```
the test passes.  If I restore line 737 to the original as follows:
```
    DSTEQR_F77(CHAR_MACRO(COMPZ), &n, D, E, Z, &ldz, WORK, info);
```
then the test fails.  However, if I do the following, to emulate the other Fortran 77 string ABI thing:
```
    const unsigned int ONE = 1;
    DSTEQR_F77(&COMPZ, ONE, &n, D, E, Z, &ldz, WORK, info);
```
the test segfaults right away, without printing the above error message.

**Comment by mhoemmen** [comment]
@bathmatt @hkthorn It looks like this is an issue relating to the IBM (Open)POWER Fortran ABI.  I am gathering info from @nmhamster on how to proceed.

**Comment by mhoemmen** [comment]
btw this error implies that calling ANY `Teuchos::BLAS` or `Teuchos::LAPACK` function that takes a `const char` argument could be badly messed up.  We're only lucky that Tpetra tends not to call the BLAS except for GEMM.

**Comment by mhoemmen** [comment]
@bathmatt Here are the packages that need fixing (grep for `CHAR_MACRO` or `_fcd`, exclude `ex_fcdcpy` in Seacas):
  - Amesos (ScaLAPACK wrappers)
  - Anasazi (modal analysis solvers)
  - AztecOO
  - Epetra
  - ML
  - Teuchos
  - Tpetra

That's a lot.  It might not be _hard_ to fix (that's the value of the macros), but if the proposed fix breaks AztecOO or ML, it could cause a lot of trouble.

I need input from @trilinos/framework before I proceed.  I consulted with the primary internal application's DevOps team, and they said that any solution that does CMake detection, needs to have a manual override option, in order to support cross compilation.

**Comment by mhoemmen** [comment]
@trilinos/anasazi @trilinos/amesos @trilinos/teuchos @trilinos/muelu @trilinos/tpetra

**Comment by mhoemmen** [comment]
@bathmatt Per discussion in #1212 and #1208, this is a general Trilinos issue that might benefit from an outsourced solution.  However, it appears to be tied to use of `gfortran` to build the BLAS and LAPACK.  Would you consider using ESSL instead as a work-around?

**Comment by bathmatt** [comment]
can you send me the changes to my configure script to test it?  I'm o.k.
with trying it.

**Comment by mhoemmen** [comment]
@bathmatt I haven't actually tried ESSL yet.  Let me poke around a bit.

**Comment by nmhamster** [comment]
@bathmatt @mhoemmen - please remember ESSL does not implement all of the BLAS or LAPACK functions, so you end up needing a fall back library.

**Comment by mhoemmen** [comment]
Ride has the `ibm/essl/5.4.0` module, but I'll have to dig a bit to figure out how to link against it.

**Comment by mhoemmen** [comment]
```
$ nm /home/projects/pwr8-rhel73-lsf/ibm/essl/5.4.0/lib/libessl.so | grep dgemm

...
00000000000ac380 T cblas_dgemm
00000000009c3120 T dgemm
00000000009c416c T dgemm_
...

$ nm /home/projects/pwr8-rhel73-lsf/ibm/essl/5.4.0/lib/libessl.so | grep dsteqr

0000000000d936e0 t dsteqr
```
That looks promising... let me see if this is the right library with respect to OpenMP.  I'm also not sure if the IBM OpenMP run-time environment will fight with GCC's OpenMP.  Let me read the ESSL documentation a bit.

**Comment by mhoemmen** [comment]
I looked at `/home/projects/pwr8-rhel73-lsf/ibm/essl/5.4.0/include/essl.h` -- it has the equivalent of `extern "C"` declarations that show that the library represents `character(*), intent(in)` just as `const char*`.  Thus, if we can actually use ESSL, it should solve your problem.

**Comment by mhoemmen** [comment]
@bathmatt hm, ESSL documentation says that GCC OpenMP run-time is not compatible with XL OpenMP, that ESSL uses.  Thus, you'll have to run ESSL in sequential mode if using it with GCC.  Is that bad?

**Comment by mhoemmen** [comment]
Here are the necessary linker flags for doing so:
```
gcc ${FILENAME} -lessl -lxlf90_r -lxlfmath -lm \
  -L/opt/ibm/xlsmp/${xlsmp_version.release}/lib \
  -L/opt/ibm/xlf/${xlf_version.release}/lib \
  -R/opt/ibm/lib
```

**Comment by bathmatt** [comment]
I don't understand your comment @mhoemmen we need to test with openmp and kokkos, are you saying we cannot compile with openmp?  or are you saying we cannot use openmp in blas routines?

**Comment by bathmatt** [comment]
BTW, have you tested essl yet?

**Comment by mhoemmen** [comment]
@bathmatt wrote:
> or are you saying we cannot use openmp in blas routines?

This.  You can use GCC's OpenMP in your code, but you can't use GCC + OpenMP with OpenMP happening in the ESSL routines.  If you want ESSL to use OpenMP, you have to build with the XL compiler.

**Comment by mhoemmen** [comment]
@bathmatt If the terms in my above comment are acceptable, I'll test ESSL.  Otherwise, if you really need OpenMP in your BLAS + GCC with OpenMP, you'll need a true fix for all the linkage issues mentioned above.

**Comment by bathmatt** [comment]
As long as I can have parallel_fors in my code, I'm fine w/ this fix

**Comment by mhoemmen** [comment]
@bathmatt Cool, I'll try ESSL.  That's the native BLAS + LAPACK anyway, and the logical choice with the XL compiler (which is supposed to be preferred on this platform -- we should be disappointed if gcc ends up better).

**Comment by bathmatt** [comment]
@mhoemmen did you run with essl?  Did it resolve the issue?  What are the configure lines?

**Comment by mhoemmen** [comment]
@bathmatt yo, i actually on occasion have other work to do so don't rush me :-)

**Comment by mhoemmen** [comment]
It might take a while, since we had to delete a bunch of files on ride in order to clear out the underwater disk.

**Comment by bathmatt** [comment]
I can add it to my testing, just not sure on the exact line I need to add..

**Comment by mhoemmen** [comment]
I'm trying something out but it's not quite working.  I think you can just go into that script you sent me and change LAPACK_LIB to the above line, but I'm having some troubles with that.

**Comment by mhoemmen** [comment]
I got the build to work by changing LAPACK_LIB in your ride script to the above line, but none of the tests are enabled -- I'll have to try it again.

**Comment by mhoemmen** [comment]
Argh, the script set `Teuchos_ENABLE_LONG_LONG_INT=OFF` again -- please don't do that :(

**Comment by mhoemmen** [comment]
Let's see, Tpetra tests passed.  Let me try Belos tests.

**Comment by mhoemmen** [comment]
Argh, hold on, had the wrong build directory.  Let me try again.

**Comment by mhoemmen** [comment]
OK, that didn't work -- link errors.  Let me see if the essl module got loaded.

**Comment by mhoemmen** [comment]
I loaded the essl module (`ibm/essl/5.4.0`), but still got this:
```
/usr/bin/ld: cannot find -lxlf90_r
/usr/bin/ld: cannot find -lxlfmath
```
We have to dig into this a bit more.

**Comment by nmhamster** [comment]
@mhoemmen - you also need the IBM compilers loaded for those even if you are using GCC. Load IBM first and then GCC

**Comment by mhoemmen** [comment]
@nmhamster Thanks!  Let me try that and report back.

**Comment by mhoemmen** [comment]
Hm, same message, let me tweak the link lines a bit first...

**Comment by mhoemmen** [comment]
I can't get this to work.  Here's my `LD_LIBRARY_PATH`:
```
/home/projects/pwr8-rhel73-lsf/openssl/1.0.2j/lib:/home/projects/pwr8-rhel73-lsf/perl/5.22.1/lib:/home/projects/pwr8-rhel73-lsf/tcl/8.5.19/lib:/home/projects/pwr8-rhel73-lsf/curl/7.47.1/lib:/home/projects/pwr8-rhel73-lsf/yamlcpp/20170104/lib:/home/projects/pwr8-rhel73-lsf/parmetis/4.0.3/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/metis/5.0.1/gcc/5.4.0/lib:/home/projects/pwr8-rhel73-lsf/boost/1.60.0/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/superlu/4.3.0/gcc/5.4.0/lib:/home/projects/pwr8-rhel73-lsf/openblas/0.2.19/gcc/5.3.0/lib:/home/projects/pwr8-rhel73-lsf/cgns/3.2.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/netcdf/4.4.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/hdf5/1.8.17/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/pnetcdf/1.6.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44/lib:/home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib64:/home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib:/home/projects/pwr8-rhel73-lsf/gcc/5.4.0/lib/gcc/powerpc64le-unknown-linux-gnu/5.4.0:/home/projects/pwr8-rhel73-lsf/cloog/0.18.4/lib:/home/projects/pwr8-rhel73-lsf/isl/0.17.1/lib:/home/projects/pwr8-rhel73-lsf/gmp/6.1.0/lib:/home/projects/pwr8-rhel73-lsf/mpc/1.0.3/lib:/home/projects/pwr8-rhel73-lsf/mpfr/3.1.5/lib:/home/projects/pwr8-rhel73-lsf/zlib/1.2.8/lib:/home/projects/pwr8-rhel73-lsf/papi/5.5.0/lib:/home/projects/pwr8-rhel73-lsf/numa/2.0.11/lib:/home/projects/pwr8-rhel73-lsf/cuda/8.0.44/lib/nvidia:/home/projects/pwr8-rhel73-lsf/cuda/8.0.44/lib64:/home/projects/pwr8-rhel73-lsf/ibm/essl/5.4.0/lib:/home/projects/pwr8-rhel73-lsf/ibm/xl/xlf/15.1.5/lib64:/home/projects/pwr8-rhel73-lsf/ibm/xl/xlf/15.1.5/lib:/home/projects/pwr8-rhel73-lsf/ibm/xl/xlC/13.1.5/lib64:/home/projects/pwr8-rhel73-lsf/ibm/xl/xlC/13.1.5/lib:/home/projects/pwr8-rhel73-lsf/seacas/20170224/openmpi/none/gcc/5.4.0/cuda/none/lib:/opt/lsf/10.1/linux3.10-glibc2.17-ppc64le/lib
```
and here's my list of loaded modules:
```
Currently Loaded Modulefiles:
  1) seacas/20170224/openmpi/none/gcc/5.4.0/cuda/none     12) cloog/0.18.4                                         23) metis/5.0.1/gcc/5.4.0
  2) ibm/xl/13.1.5                                        13) gcc/5.4.0                                            24) parmetis/4.0.3/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
  3) ibm/essl/5.4.0                                       14) java/ibm/sdk/8.0.0                                   25) yaml-cpp/20170104
  4) cuda/8.0.44                                          15) openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44                 26) curl/7.47.1
  5) numa/2.0.11                                          16) pnetcdf/1.6.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44   27) cmake/3.6.2
  6) papi/5.5.0                                           17) hdf5/1.8.17/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44     28) tcl/8.5.19
  7) zlib/1.2.8                                           18) netcdf/4.4.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44    29) perl/5.22.1
  8) mpfr/3.1.5                                           19) cgns/3.2.1/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44      30) openssl/1.0.2j
  9) mpc/1.0.3                                            20) openblas/0.2.19/gcc/5.4.0                            31) git/2.10.1
 10) gmp/6.1.0                                            21) superlu/4.3.0/gcc/5.4.0                              32) valgrind/3.11.0
 11) isl/0.17.1                                           22) boost/1.60.0/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44    33) devpack/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.44
```

**Comment by jhux2** [comment]
> I can't get this to work. Here's my LD_LIBRARY_PATH:

@mhoemmen Did you get Trilinos to configure and build on ride?

**Comment by mhoemmen** [comment]
@jhux2 I couldn't figure out how to get Trilinos to configure on ride, with gcc and ESSL.

**Comment by mhoemmen** [comment]
@jhux2 Hold on, please see discussion here: 

https://github.com/trilinos/Trilinos/issues/1208#issuecomment-293115514

**Comment by nmhamster** [comment]
@mhoemmen - can you send me the errors/warnings you were getting please? I think we can work through this and fix it.

**Comment by jhux2** [comment]
>> I can't get this to work. Here's my LD_LIBRARY_PATH:
>> 
> @mhoemmen Did you get Trilinos to configure and build on ride?

@mhoemmen @bathmatt's [configure script](https://github.com/trilinos/Trilinos/issues/1191#issuecomment-291571944) worked fine for me.  My compile completed without error, as well.

**Comment by mhoemmen** [comment]
@jhux2 The problem manifests as run-time errors, not build or configure errors.  The issue is that Trilinos doesn't use the right ABI for LAPACK functions.

**Comment by mhoemmen** [comment]
@jhux2 Matt's script configures and builds fine for me too :-)  The issue is that I'm trying to change it to use ESSL, and that doesn't work.  I suspect it's because of some things @nmhamster is saying over in #1208, namely that ESSL is not a complete BLAS / LAPACK implementation, so you have to link with OpenBLAS as well.

**Comment by nmhamster** [comment]
@jhux2 can you post your errors please? Are you on white or ride?

**Comment by maherou** [comment]
Jim Willenbring has asked Tech-X about taking on this responsibility.  Kris Beckwith has requested and received access to ride.  James Elliott is also engaged. The issue may be related to configuration issues.
ACTION: @jwillenbring will follow up with @krcb

**Comment by bathmatt** [comment]
So, I skimmed through the issue, I assume there is no work around at this point? ?  Should we assume that Trilinos on Ride is not viable for packages that use BLAS?

**Comment by krcb** [comment]
We've been updating #1208 as with the status of our investigation. As yet, we don't have a work around; investigations so are pointing towards how the different libraries are linked.

**Comment by bartlettroscoe** [comment]
> So, I skimmed through the issue, I assume there is no work around at this point? ? Should we assume that Trilinos on Ride is not viable for packages that use BLAS?

Can I ask a dumb question?  To get past this issue, at least in the short-term, can we not just build the reference BLAS and LAPACK with the supplied compilers and get things working for now?  Then, offline, someone who cares about a higher-performance BLAS implementation can work on the side to get these issues worked out.

**Comment by krcb** [comment]
@bartlettroscoe I agree completely. AFAIK, nobody has tried simply compiling Trillinos against only OpenBLAS (no ESSL) on the machine as yet. This was suggested on #1208 previously and is my first step when I get onto Ride.

**Comment by bartlettroscoe** [comment]
@bathmatt and @rppawlo,

FYI:

I just merge the Trilinos PR # 1284 to the ‘develop’ branch last night.  Therefore, you should hopefully not see these LAPACK errors on 'ride' you were seeing before.  Hopefully this will allow some of your Panzer, Drekar, and EMPIRE tests to now pass that were failing before.

However, note that there are other Trilinos test failures (segfaults) on 'ride' that are likely **NOT** related to these LAPACK calling errors as summarized in https://github.com/trilinos/Trilinos/pull/1284#issuecomment-300589150 .  Therefore, someone is going to need to dig into those as a separate issue (i.e. a new GitHub issue that blocks this issue that focuses on those segfaults).

I will update the "Next Action Status" field for the current status of this.

**Comment by rppawlo** [comment]
just ran with the updates and it did not fix the anasazi solver on ride. Still seeing a seg fault.

**Comment by bartlettroscoe** [comment]
@rppawlo:

> just ran with the updates and it did not fix the anasazi solver on ride. Still seeing a seg fault.

But the LAPACK errors are gone?  That is all that the merged PR #1284 sought to address.

As I commented above, there are other problems on 'ride' that may not have anything to do with mixed language calls to LAPACK routines.  As of now, I don't know of anyone looking deeply into these segfaults.  Perhaps we can ask the new Tech-X contractor @kruger to take a look at this?

@kruger, given the importance of the problems on 'ride', can you consider talking with @krcb about the status of 'ride' and consider a deeper dive to try to get to the bottom of what is happening on this machine?  That is not originally the type of task that I imagined you starting with but since I suspect this is a system issue on an ATDM platform, it might fall under the scrope of the ATDM Tools & Dev Env project.

**Comment by krcb** [comment]
@bartlettroscoe I discussed this with @kruger and have got him setup on 'ride'. For my own insight, I just built Belos with ```-DTrilinos_ENABLE_DEBUG=ON``` and  ```-DCMAKE_BUILD_TYPE=Debug```. All but one of the Belos tests pass; the failing one is that above ```Belos_Tpetra_PseudoBlockCG_hb_test.exe```. We still have a segfault on:

```
#0  0x00003fffffffbce8 in ?? ()
#1  0x00000000105b3da0 in Belos::PseudoBlockCGSolMgr<double, Tpetra::MultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace>, false>, Tpetra::Operator<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >, true>::solve (this=0x0) at /ascldap/users/kbeckwi/Trilinos/packages/belos/src/BelosPseudoBlockCGSolMgr.hpp:988
#2  0x00003fff00000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

So, while it appears that the LAPACK error is going, there is still a segfault. Note that if I deactivate Debug, then other tests fail, but this test passes. @kruger will investigate further.

**Comment by kruger** [comment]
In progress.   

Although the Teuchos failures and `Belos_Tpetra_PseudoBlockCG_hb_test.exe ` failure are with `STEQR` tests like previous issues, it does not appear to be a fortran interop issue per se.  The fortran code is getting the values OK, and the initial call back is OK.  It is when it returns from `LAPACK.STEQR` that the seg fault occurs.   

I've built my own openblas just to ensure it is not a bad module build.  I've build a version of STEQR as part of the Teuchos build tree.  They all give the same thing.

I suspect that it has to do with the NDEBUG macro not being defined, but since I don't see it in Teuchos, this is a very uncertain hypothesis.  It could still be memory corruption beteween the C++ and fortran, but valgrind isn't telling me anything useful, and just print lines are not showing anything.

**Comment by bartlettroscoe** [comment]
I want to add a comment here that some of the more recent experimentation by @kruger mentioned in https://github.com/trilinos/Trilinos/issues/1208#issuecomment-303165893 shows that the PR #1284 that was accepted may not, in fact, fix the problem with calling Fortran routines from C++ on this machine with GCC/gfortran. 

The conversation will move back to #1208 while we attack the problem further there.

**Comment by mhoemmen** [comment]
We resolved in the SART meeting this morning to subsume this issue into #1208.  Please refer to discussion there.  Thanks all!
